### PR TITLE
Refactor spectrum_vector code, fix imports, and upgrade dependencies

### PR DESCRIPTION
**Pull request by rukubrakov**
Dependencies (pyproject.toml)

Added depthcharge-ms ≥0.4.9 and metabo-depthcharge (pinned to GitHub main)
Replaced local editable path for metabo-depthcharge with git+https://github.com/bittremieuxlab/metabo-depthcharge.git
Dead code removal

spectrum.py: stripped leftover spectrum_vector, mz_array, intensity_array, max_peak fields and the messy old __getstate__/__setstate__
encoding.py: deleted encode_adduct_mass, OneHotEncoding, ADDUCT_TO_MASS, import pandas — only ion activation/method helpers remain
preprocessor.py: removed large block of unused spectrum-vector preprocessing code
train_utils.py: removed ~12 lines of dead utility code
weighted_sampling.py: minor cleanup
ILP crash fix (edit_distance.py)

catch_errors=False → True; distance == -1 (solver failure) now maps to np.nan instead of crashing the preprocessing run
Adduct encoding (encoder_dataset_builder.py, multitask_dataset_builder.py, multitask_dataset.py)

Replaced one-hot adduct encoding over 30+ adducts with categorical integer index via encode_adduct() from metabo-depthcharge (9 known adducts + 0=unknown)
Adduct arrays changed from float32 to int64; similarity_models.py uses .long() instead of .float()
CE encoding (encoder_dataset_builder.py, multitask_dataset_builder.py)

Replaced raw float passthrough with encode_collision_energy() from metabo-depthcharge, which normalises CE÷100 and handles None/NaN/stepped strings
MGF loader (loaders.py)

CE field now read with fallback chain: collision_energy → collision_energy_1 → ce, covering MSG, Spectraverse and legacy formats
Spectrum encoder — full rewrite (spectrum_encoder.py, similarity_models.py)

Old: subclassed depthcharge's SpectrumTransformerEncoder, concatenate+project peak encoding, raw scalar global token, one-hot adduct via FloatEncoder
New: subclasses metabo-depthcharge's SpectrumEncoder — additive sinusoidal peak encoding, learned CLS + sinusoidal precursor_mz global token, MetadataEncoder for adduct (embedding) and CE (linear + zero-mask), pool="attention" (AttnAggregator over all tokens)
Zero imports from depthcharge in the encoder; ion_mode/activation/method use nn.Linear and are injected via global_token_hook override
similarity_models.py: removed 3 tuple-unpack (emb, _) sites and 3 [:, 0, :] slices — encoder now returns (B, d_model) directly
SLURM scripts (new)

train_joint_metadata.slurm.sh: trains on joint dataset with use_adduct, use_ce, use_ion_mode
inference_joint_metadata.slurm.sh: inference with the attn-pool model, proper paths.output_dir and model feature flags
inference_joint_metadata_cls.slurm.sh: inference script for the older CLS-pool checkpoint

## Summary by CodeRabbit

* **New Features**
  * Added joint metadata support (adduct, collision energy, ion activation/method/ion mode) for training and inference.
  * Integrated Classyfire-based molecular class annotations into preprocessing.
* **Improvements**
  * Updated metadata-aware spectrum encoding and dataset features to align with the new joint-metadata schema.
  * Added configurable attention pooling for model encoding; updated model loading and embedding extraction.
  * Improved spectrum parsing for SMILES and collision energy extraction.
* **Bug Fixes**
  * More robust similarity scoring when the MCES solver fails.
* **Chores**
  * Updated dependency constraints and strengthened SLURM scripts with stricter failure handling.

### CODE REVIEW

File: simba/core/data/datasets/encoder_dataset_builder.py [hunk 1]
```diff
@@ -57,32 +55,27 @@ def prepare_encoder_dataset(spectra, max_num_peaks=100):
...
-        if hasattr(spectrum, 'ce') and spectrum.ce is not None and spectrum.ce != "None":
-            ce[i] = spectrum.ce
-        else:
-            ce[i] = 0
-        
+        ce[i] = encode_collision_energy(getattr(spectrum, "ce", None))
+
         # Ion Activation
         if hasattr(spectrum, 'ion_activation') and spectrum.ion_activation is not None and spectrum.ion_activation != "None":
             ia[i] = encode_ion_activation(spectrum.ion_activation)
```
Comment (gdewael) [inline simba/core/data/datasets/encoder_dataset_builder.py]: We could also consider adding these to metabo-depthcharge to provide a unified interface to encoding metadata. Similar for `encode_ionization_method` at L81

File: simba/core/data/loaders/loaders.py [hunk 2]
```diff
@@ -458,11 +458,13 @@ def _parse_spectrum(
         else:
             adduct = None
 
-        ce = params["ce"] if "ce" in params else None
-        ia = params["ion_activation"] if "ion_activation" in params else None
-        im = params["ionization_method"] if "ionization_method" in params else None
+        # Try common MGF key variants: COLLISION_ENERGY (MSG/MassSpecGym),
+        # COLLISION_ENERGY_1 (Spectraverse stepped CE), CE (legacy)
+        ce = params.get("collision_energy") or params.get("collision_energy_1") or params.get("ce")
```
Comment (gdewael) [inline simba/core/data/loaders/loaders.py]: Could also use https://metabo-depthcharge.readthedocs.io/en/latest/api/generated/metabo_depthcharge.spec.preprocessing.CollapseSteppedCE.html But this somewhat requires to use the `Spectrum` class object defined by `metabo-depthcharge`.

Another note here: In massspecgym, normalized collision energies are stored under the "COLLISION_ENERGY" MGF field, whereas in SpectraVerse both NORMALIZED_COLLISION_ENERGY_{1,2,3} and COLLISION_ENERGY_{1,2,3} are provided. For parity, normalized collision energies should be used. (the above linked class allows this seamlessly).

File: simba/core/models/spectrum_encoder.py [hunk 3]
```diff
@@ -16,141 +44,100 @@ def __init__(
...
+        if use_adduct:
+            metadata_fields.append("adduct")
+        if use_ce:
+            metadata_fields.append("collision_energy")
+        metadata_enc = (
+            MetadataEncoder(d_model, metadata_fields) if metadata_fields else None
+        )
+
+        super().__init__(
+            *args, pool="attention", metadata_encoder=metadata_enc, **kwargs
```
Comment (gdewael) [inline simba/core/models/spectrum_encoder.py]: Note that this may have unforeseen consequences on how easy it is for the model to make its predictions conditional on metadata: pool="cls" will take the embedding of the global_token at output, where all metadata was injected.
It is not crazy to think that it easier for the model to use metadata information when explicitly selecting this token as the "summary" embedding, instead of using attention pooling (even after all the transformer layers).

Fine for now, but mental note in future experiments on metadata inclusion, relevant to discuss with Janne and Sebastian, as they have been heavily experimenting with this.

File: tools/slurm/inference_joint_metadata_cls.slurm.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+#SBATCH --cpus-per-task=4
+#SBATCH --mem=64G
+#SBATCH -o /home/nkubrakov/simba-integration/logs/inference_joint_metadata_cls_%j.out
+#SBATCH -e /home/nkubrakov/simba-integration/logs/inference_joint_metadata_cls_%j.err
+
+# Inference on joint dataset using model trained with adduct + CE + ion_mode metadata,
+# CLS-pool encoder (metadata_adduct_ce_ionmode checkpoint).
+# NOTE: current code uses pool="attention"; the AttnAggregator weights will be randomly
+# initialized since this checkpoint was trained with pool=None. Transformer/metadata
+# encoder weights load correctly. Results are approximate.
```
[the remark this answers is not in the document]
Comment (gdewael) [reply tools/slurm/inference_joint_metadata_cls.slurm.sh:16]: Please check this as this could impact the comparison you shared earlier

**Review by gdewael — COMMENTED** [review]
I've left some comments/thoughts here and there. I dont have other major notes on how the code should be restructured to utilize metabo-depthcharge.
